### ISSUE TITLE
Depth-colour projection / new 3D rendering mode

### ISSUE THREAD

**Issue by haesleinhuepf** [napari/napari#4300]
## 🚀 Feature
Hi all,

## Motivation

a colleague of mine asked if a depth-colour projection could be a nice napari-plugin. I'd say it would be better suited as rendering mode in napari core. If the colour would change when rotating the sample, this could give a visual impression we have never seen in other tools such as Fiji before.

## Pitch

The depth colour projection (called "temporal colour code" in Fiji) is used to visualize the position of bright intensity in a Z-stack. E.g. in this image ([source](https://twitter.com/PavelTomancak/status/1505853968794140675?s=20&t=Rkv3d4tHD78lrvakc976dQ)) blue pixels are in the front, red pixels are in the back of the embryo:
![image](https://user-images.githubusercontent.com/12660498/159649441-4be86f20-1cde-47a4-b91f-efa0b16cf3e9.png)

## Alternatives

It could be implemented as image processing operation producing a static projection. This functionality could be identical compared to Fiji. I think though, the interactive rendering would be much more insightful.

## Additional context

The [implementation in Fiji, written in macro](https://github.com/fiji/fiji/blob/master/plugins/Scripts/Image/Hyperstacks/Temporal-Color_Code.ijm) by Kota @miura, might be inspiring.

**Comment by miura** [comment]
I feel responsible for this migration...

**Comment by haesleinhuepf** [comment]
Great to hear that Kota! I'd hope that maybe @alisterburt @brisvag and/or @kevinyamauchi could help, because they implemented similar "shaders" earlier.

**Comment by psobolewskiPhD** [comment]
This kind of visualization, where z-position is indicated by color, is also very nice for some materials science applications, where surface topography or roughness are presented.
I can also imaging it being useful in cartography and related imaging...

**Comment by kevinyamauchi** [comment]
Hey everyone. This actually feels very doable as a vispy shader filter now that vispy is setting the depth buffer for most image rendering modes (final ones contributed by @alisterburt in https://github.com/vispy/vispy/pull/2305!).

Basically, the shader filter would have to set the `gl_FragColor` by the value in `gl_FragDepth`. The tricky thing is setting the limits for mapping depth to color, but maybe that's just user settable (like contrast limits).

I think this shader filter is basically what is needed. I am not sure if it's working though (I recall some issues with it in the past).

https://github.com/vispy/vispy/blob/30fd4b9df5f83eadbc75abade82aa42c875d2d96/vispy/visuals/filters/color.py#L154-L181

**Comment by kevinyamauchi** [comment]
I don't have time to take this on right now, but happy to chat if anybody would like to give it a go!

**Comment by alisterburt** [comment]
thanks for bringing this up @haesleinhuepf - such a great suggestion!

@kevinyamauchi reading that shader filter I'm not sure what's going on but I agree this is super achievable and not a crazy amount of work, the hardest part will be exposing it in napari in a useful way. We would have to rework the contrast limits to account for the possible depth values (and decide if that depth should be set in data/scene space).

We definitely need to use the depth *after* transforming into world space to avoid similar issues to #3646 🙂 

I'm in a similar boat to Kevin, not high priority for me but would be excited to help someone work through it no matter their familiarity with vispy/shaders!

**Comment by miura** [comment]
I'm very much encouraged by your comments, @kevinyamauchi , @alisterburt , @psobolewskiPhD  , and @haesleinhuepf OK, I assign myself to explore my first napari plugin if I can get help from you guys guiding me to the world. It will be a great chance for me to learn using napari API.

**Comment by brisvag** [comment]
I played around with this a bit yesterday; It's a very cool idea, and it should be doable. These are the issues I found:

- Instead of simply using a `position` from the vertex shader, one needs to read the `gl_FragDepth` which is set at the end of the fragment shader, because depending on the rendering mode this will change
- Using the raw number of `gl_FragDepth` won't work, because its scale is crazy and squished; instead, we need to find a way too convert this into a number relative to the closest and furthesy point of the *volume*, rather than the whole scene. This is the hard bit I think...

**Comment by alisterburt** [comment]
Great to have you on board @miura ! This won't be the simplest introduction to napari, it's quite in the guts but that's fun 😄 

@brisvag is right, the value in the depth buffer (`gl_FragDepth`) is [not linear](https://learnopengl.com/Advanced-OpenGL/Depth-testing)
<img width="558" alt="image" src="https://user-images.githubusercontent.com/7307488/159934216-2aaf0e90-8de4-4123-a60e-6fa12579c599.png">

the same article also shows how to linearise the value from the depth buffer and scale it back into the range [0,1] though! 🥳 

```glsl
#version 330 core
out vec4 FragColor;

float near = 0.1; 
float far  = 100.0; 
  
float LinearizeDepth(float depth) 
{
    float z = depth * 2.0 - 1.0; // back to NDC 
    return (2.0 * near * far) / (far + near - z * (far - near));	
}

void main()
{             
    float depth = LinearizeDepth(gl_FragCoord.z) / far; // divide by far for demonstration
    FragColor = vec4(vec3(depth), 1.0);
}
```

> Because the linearized depth values range from near to far most of its values will be above 1.0 and displayed as completely white. By dividing the linear depth value by far in the main function we convert the linear depth value to the range [0, 1].

Our near and far clipping planes seem to be fixed at z=1 and -1
https://github.com/vispy/vispy/blob/30fd4b9df5f83eadbc75abade82aa42c875d2d96/vispy/visuals/volume.py#L68-L74

So a little shader filter like the one @kevinyamauchi linked to above which runs at the end of the volume visual would have to
- calculates a linearised `depth` value from `gl_FragDepth` 
- puts that `depth` value in the range [0, 1] as above
- applies the colormap to the value (`gl_FragColor = applyColormap(depth)`)

should be what we need! @miura don't hesitate if more guidance than this would be useful, maybe @brisvag could provide a simple example of attaching/detaching a shader filter?

**Comment by brisvag** [comment]
> the same article also shows how to linearise the value from the depth buffer and scale it back into the range [0,1] though! partying_face

Great!

However, there's another complication: if one only uses the fragment depth, there is no difference between bright pixels and dark ones; so you want to actually still account for the original brightness. However, you cannot simply sample the texture, because there's some preprocessing to do (projection and whatnot). So you need to take the output of the first colormap application... so one could do something like:

```glsl
intensity = length(gl_FragColor);
```

but I'm not sure there's a way to easily convert *any* colormap output to a reasonable value for intensity. Actually, the more I talk about this, and the more I think this should just be a different rendering method altogether.

**Comment by haesleinhuepf** [comment]
> but I'm not sure there's a way to easily convert _any_ colormap output to a reasonable value for intensity

The answer may lie in these lines in the original implementation by @miura : 
https://github.com/fiji/fiji/blob/master/plugins/Scripts/Image/Hyperstacks/Temporal-Color_Code.ijm#L73-L88

**Comment by arteys** [comment]
As far as I understand there is nothing new since the last post? Faced with the need to do this very thing in napari....

**Comment by jni** [comment]
Yeah, about a year ago @royerloic paired with @brisvag for a bit but had issues getting the implementation correct, and I think the effort lost steam. It would be good to get the latest status/code from those two posted here so someone can pick up the thread maybe... 🙏

**Comment by brisvag** [comment]
There's a completely new (and relatively active) effort from @aganders3 over here: https://github.com/vispy/vispy/pull/2550 (you commented, @jni xD).

This PR would allow this and much more! It needs a bit more love I think, but it's not far.

**Comment by jni** [comment]
Oops! 😂 Well... Great! 😂

**Comment by jichen12** [comment]
Hi @arteys , 

I found a plug in in napari [https://www.napari-hub.org/plugins/napari-zplane-depth-colorizer](url). I hope it helps.
